app/sentinel/internal/safe/post.go

```diff
@@ -156,31 +157,84 @@ func doPost(client *http.Client, p string, md []byte) {
 	respond(r)
 }
 
-func Post(workloadId, secret, namespace, backingStore string, useKubernetes bool,
-	template string, format string, encrypt, deleteSecret, appendSecret bool,
-	inputKeys string,
+func Post(parentContext context.Context, workloadId, secret, namespace, backingStore string,
+	useKubernetes bool, template string, format string, encrypt, deleteSecret, appendSecret bool, inputKeys string,
 ) {
-	ctx, cancel := context.WithCancel(context.Background())
+	ctxWithTimeout, cancel := context.WithTimeout(
+		parentContext,
+		env.SafeSourceAcquisitionTimeout(),
+	)
 	defer cancel()
 
-	source, proceed := acquireSource(ctx)
-	defer func() {
-		if source == nil {
+	sourceChan := make(chan *workloadapi.X509Source)
+	proceedChan := make(chan bool)
+
+	go func() {
+		source, proceed := acquireSource(ctxWithTimeout)
+		sourceChan <- source
+		proceedChan <- proceed
+	}()
+
+	select {
+	case <-ctxWithTimeout.Done():
+		if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+			fmt.Println("Post: I cannot execute command because I cannot talk to SPIRE.")
+			fmt.Println("")
 			return
 		}
-		err := source.Close()
-		if err != nil {
-			log.Println("Problem closing the workload source.")
+
+		fmt.Println("Post: Operation was cancelled due to an unknown reason.")
+	case source := <-sourceChan:
+		defer func() {
```

Comment:
From this point, it’s essentially moving everything under this `case`.

We can, again, move the logic into its own function.

I just want to change as little as possible until I am sure that things work as expected.